### PR TITLE
fix(rollback): Ensure that rollback stages are injected correctly

### PR DESCRIPTION
**Pull request by ajordens**
This PR handles generating a `refId` for a synthetic stage that has
pre-existing siblings.

In such cases, the `refId` should be unique and offset by the number of
siblings.

All `onFailureStages()` should be run linearly otherwise you run the
risk of a parent stage being notified prematurely.

### CODE REVIEW

**Comment by ajordens** [comment]
image

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/CreateServerGroupStage.groovy [hunk 1]
```diff
@@ -106,7 +106,6 @@ class CreateServerGroupStage extends AbstractDeployStrategyStage {
     }
 
     if (strategySupportsRollback) {
-      // TODO: should this append after any existing stages in the graph?
```
Comment (ajordens) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/CreateServerGroupStage.groovy]: The answer was yes ... I didn't see anything on `StageGraphBuilder` that would let you `connect()` to the last added node.

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/graph/StageGraphBuilder.java [hunk 2]
```diff
@@ -148,11 +148,18 @@ public void connect(@Nonnull Stage previous, @Nonnull Stage next) {
   }
 
   private String generateRefId() {
+    long offset = parent
```
Comment (ajordens) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/graph/StageGraphBuilder.java]: If we're adding stages to an already built execution, it's important to not re-use a `refId`.

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/graph/StageGraphBuilder.java [hunk 3]
```diff
@@ -148,11 +148,18 @@ public void connect(@Nonnull Stage previous, @Nonnull Stage next) {
...
+      .getExecution()
+      .getStages()
+      .stream()
+      .filter(i -> parent.getId().equals(i.getParentStageId()) && type == i.getSyntheticStageOwner())
```
Comment (ajordens) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/graph/StageGraphBuilder.java]: not necessary the most elegant, alternatives?

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/CompleteStageHandler.kt [hunk 4]
```diff
@@ -165,7 +165,14 @@ class CompleteStageHandler(
 
     val graph = StageGraphBuilder.afterStages(this)
     builder().onFailureStages(this, graph)
+
     val onFailureStages = graph.build().toList()
+    onFailureStages.forEachIndexed { index, stage ->
+      if (index > 0) {
+        // all on failure stages should be run linearly
+        graph.connect(onFailureStages.get(index - 1), stage)
```
Comment (ajordens) [inline orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/CompleteStageHandler.kt]: Without this, we run the risk of one of the stages completing and notifying it's parent ... and completing the execution with another stage still running in parallel.

My expectation is that > 1 onFailure stages will run linearly (they did previously).

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/CreateServerGroupStage.groovy (hunk 1 above)
Comment (robfletcher) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/CreateServerGroupStage.groovy]: Yeah, I think `StageGraphBuilder` could be a little easier to use for certain common scenarios.

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/graph/StageGraphBuilder.java (hunk 3 above)
Comment (robfletcher) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/graph/StageGraphBuilder.java]: Seems reasonable. I can't think of a neater way off the top of my head.

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/CompleteStageHandler.kt (hunk 4 above)
Comment (robfletcher) [reply orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/CompleteStageHandler.kt]: I'd like to address the root issue as well (not in this PR). The execution should not complete if anything is still in a `RUNNING` state and if it does that's a bug.

**Pull request merged**